relay/relay-network.md

```diff
@@ -0,0 +1,67 @@
+# Relay Network
+
+Relay network is defined by the federation of relay servers exchanging messages to route from a publishing client to a subscribing client on the same topic.
+
+## Goals
+
+Relay servers will route messages through other servers in the network that also have matching subscribers on a topic whilst maintaing the same expected behavior described by a publisher when requested to its own server
+
+## Network Message
+
+An Network message is defined as a serialized message of a successfully cached message from a publishing request from a connected client with one of the servers.
+
+Context:
+
+- **topic** - a string defining the target topic for message subscription.
+- **hash** - a string defining the message id as the sha256 hash of itself.
```

Comment:
I think we should define this hash with an additional time value. I bumped into an error with redis recently with the jsonrpc tests in the ts-relay where two messages with the value of "SECOND_MESSAGE" would store in the same hash of redis in the message table.

My solution was to add an additional field to the message identifier: hash(topic, "SECOND_MESSAGE"). This ensure that each test run would make a new unique message since the topic was random.

I know that all message payloads are encrypted to therefor they are random enough and therefor we can ensure the uniqueness of each message. The problem occurs is that if the message payloads are not encrypted and are instead plain text one can send two messages with the same value and same topic and they would end up having the same `hash`.

I would trust more the output of the `hash` field if it contained some sort of salt. My suggestion is that we use `sha156(payload, jsonrp_id)` since the jsonrpcid is currently a semi random timer field. This is not a  perfect salt but at least now we can store **IF NEEDED** two identical payloads but sent at different times. 

Any thoughts on the suggestion? It is merely because I want to avoid bumping into a bug where we actually want to send two payloads identical payloads in a single topic.

Reply:
There is a risk that the network could be spammed with identical payloads but different `jsonrpcid`, but the network could still be spammed with spam/random message payloads.